{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n",
      "match\n"
     ]
    }
   ],
   "source": [
    "import csv #to import the csv library to be able to work with this type of files (it also includes .tsv)\n",
    "\n",
    "germplasmfile = open(\"Germplasm.tsv\", \"r\")#open it as a read-only file \"r\"\n",
    "locusgenefile = open(\"LocusGene.tsv\", \"r\")\n",
    "\n",
    "germplasmfile.seek(0)   \n",
    "locusgenefile.seek(0)#.seek takes the pointer to the point in the file you want, in this case, to the beginning (0)\n",
    "\n",
    "listalocus=[] #creates a new list\n",
    "i=0 #gives a value of 0 to variable i so that at each step of the loop it will change with the \"i=i+1\" that follows.\n",
    "\n",
    "for line1 in germplasmfile.readlines(): #reads each line of the file, \"line1\" is the pointer on that file\n",
    "    listalocus.append(line1.split(\"\\t\")[0]) #appends (.append) to the empty list we have created the first element ([0]) \n",
    "                                            #of each line of the file \"germplasmfile\", taking into account that the \n",
    "                                            #elements are separated by tabs (line1.split(\"\\t\"))   \n",
    "    \n",
    "for line2 in locusgenefile.readlines():\n",
    "    line2 = line2.split(\"\\t\")[0] #saves in the variable \"line2\" the first element of each line of the file \"locusgenefile\". \n",
    "                                 #In each step of the loop the line is overwritten, so that there will be only one line  \n",
    "                                 #for each step of the loop.\n",
    "\n",
    "    \n",
    "    if listalocus[i] == line2: #within this second \"for\", we make the conditional of equality between the two loops\n",
    "        print(\"match\")\n",
    "    else:\n",
    "        print(\"no match\")\n",
    "        \n",
    "    i=i+1 #at the end, we add 1 to the variable i so that it advances at the same time as the other loop, i.e. both \n",
    "          #files advance at the same time.\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "(pymysql.err.ProgrammingError) (1007, \"Can't create database 'genetics'; database exists\")\n",
      "[SQL: create database genetics;]\n",
      "(Background on this error at: http://sqlalche.me/e/f405)\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "6 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "4 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>locus</td>\n",
       "        <td>varchar(9)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "        <td>varchar(30)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>phenotype</td>\n",
       "        <td>varchar(1000)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>pubmed</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('locus', 'varchar(9)', 'NO', 'PRI', None, ''),\n",
       " ('germplasm', 'varchar(30)', 'NO', '', None, ''),\n",
       " ('phenotype', 'varchar(1000)', 'NO', '', None, ''),\n",
       " ('pubmed', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop database genetics; #we drop the database and the tables so that there are no problems creating them, \n",
    "%sql drop table germplasmtab; #in case there is another one with the same name.\n",
    "%sql drop table locusgenetab;\n",
    "%sql create database genetics;\n",
    "#%sql show databases #to check that they have been created correctly\n",
    "%sql use genetics; #to select the one I want to work with\n",
    "#%sql show tables #to check there is nothing \n",
    "%sql create table germplasmtab(locus VARCHAR(9) NOT NULL PRIMARY KEY, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(1000) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql create table locusgenetab(locus VARCHAR(9) NOT NULL PRIMARY KEY, gene VARCHAR(10) NOT NULL, proteinlengh INTEGER NOT NULL);\n",
    "%sql show tables #to check that they have been created correctly\n",
    "%sql describe germplasmtab #to see all the variables and their characteristics\n",
    "#%sql describe locusgenetab\n",
    "\n",
    "#since they have the 'locus' column in the same order, they are already matched 1:1, so it will be the primary key, instead of having an extra id column"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done!\n"
     ]
    },
    {
     "ename": "IntegrityError",
     "evalue": "(1062, \"Duplicate entry 'AT1G01040' for key 'PRIMARY'\")",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIntegrityError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-6038d9783ff8>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m                 \u001b[0;31m#print(row)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m                 \u001b[0msql\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"INSERT INTO germplasmtab(locus, germplasm, phenotype, pubmed) VALUES(%s,%s,%s,%s)\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m                 \u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m                 \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfetchall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m                 \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/cursors.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(self, query, args)\u001b[0m\n\u001b[1;32m    168\u001b[0m         \u001b[0mquery\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmogrify\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 170\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    171\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_executed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mquery\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    172\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/cursors.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self, q)\u001b[0m\n\u001b[1;32m    326\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_last_executed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m         \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    329\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_do_get_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrowcount\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(self, sql, unbuffered)\u001b[0m\n\u001b[1;32m    515\u001b[0m                 \u001b[0msql\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'surrogateescape'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    516\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_execute_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCOMMAND\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCOM_QUERY\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msql\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 517\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_affected_rows\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_query_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0munbuffered\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0munbuffered\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    518\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_affected_rows\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    519\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36m_read_query_result\u001b[0;34m(self, unbuffered)\u001b[0m\n\u001b[1;32m    730\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    731\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMySQLResult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 732\u001b[0;31m             \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    733\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_result\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    734\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_status\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1073\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1074\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1075\u001b[0;31m             \u001b[0mfirst_packet\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_packet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1076\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1077\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mfirst_packet\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_ok_packet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36m_read_packet\u001b[0;34m(self, packet_type)\u001b[0m\n\u001b[1;32m    682\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    683\u001b[0m         \u001b[0mpacket\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpacket_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuff\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 684\u001b[0;31m         \u001b[0mpacket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    685\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mpacket\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/protocol.py\u001b[0m in \u001b[0;36mcheck_error\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    218\u001b[0m             \u001b[0merrno\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_uint16\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    219\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mDEBUG\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"errno =\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrno\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 220\u001b[0;31m             \u001b[0merr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_mysql_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    221\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    222\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdump\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/err.py\u001b[0m in \u001b[0;36mraise_mysql_exception\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m    107\u001b[0m         \u001b[0merrval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'replace'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m     \u001b[0merrorclass\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror_map\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mInternalError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 109\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0merrorclass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrval\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mIntegrityError\u001b[0m: (1062, \"Duplicate entry 'AT1G01040' for key 'PRIMARY'\")"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "# Connect to the database \"genetics\", the one I have created before\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters! By the way, if I open the file with LibreOffice, UTF8 appears on it\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)# note that it is possible to delay putting changes into the database! \n",
    "                            \n",
    "try:    \n",
    "    with connection.cursor() as cursor:\n",
    "        with open (\"Germplasm.tsv\", newline= \"\\n\") as file: #open the file and declare that there is a line break and give it the name \"file\" to the variable\n",
    "            germplasmfile = csv.reader(file, delimiter='\\t', quotechar='\"') #open with the csv reader and declare that it is separated by tabs.  \n",
    "            headers= next(germplasmfile) #this is used to remove the headers so that the data in the file can be selected.\n",
    "            for row in germplasmfile:\n",
    "                #print(row)\n",
    "                sql = \"INSERT INTO germplasmtab(locus, germplasm, phenotype, pubmed) VALUES(%s,%s,%s,%s)\" #\"%s\" is referring to the data in the germplasmfile, one for each column\n",
    "                cursor.execute(sql, tuple(row)) #tuple(row) selects the data in 'row'\n",
    "                results = cursor.fetchall() \n",
    "                print(results)       \n",
    "                print(\"\")\n",
    "        \n",
    "finally:\n",
    "    print(\"done!\")\n",
    "connection.close() #is necessary so that there are no problems when opening\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351),\n",
       " ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401),\n",
       " ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125),\n",
       " ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744),\n",
       " ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956),\n",
       " ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505),\n",
       " ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900),\n",
       " ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276),\n",
       " ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837),\n",
       " ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180),\n",
       " ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444),\n",
       " ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653),\n",
       " ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278),\n",
       " ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843),\n",
       " ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770),\n",
       " ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667),\n",
       " ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018),\n",
       " ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690),\n",
       " ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066),\n",
       " ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186),\n",
       " ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662),\n",
       " ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499),\n",
       " ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832),\n",
       " ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666),\n",
       " ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749),\n",
       " ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244),\n",
       " ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216),\n",
       " ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401),\n",
       " ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251),\n",
       " ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687),\n",
       " ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678)]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasmtab;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done!\n"
     ]
    },
    {
     "ename": "IntegrityError",
     "evalue": "(1062, \"Duplicate entry 'AT1G01040' for key 'PRIMARY'\")",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIntegrityError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-b4780168d79e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m                 \u001b[0;31m#print(row)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m                 \u001b[0msql\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"INSERT INTO locusgenetab(locus, gene, proteinlengh) VALUES (%s,%s,%s)\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m                 \u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msql\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#tuple(row) selects the data in 'row'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m                 \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcursor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfetchall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m                 \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/cursors.py\u001b[0m in \u001b[0;36mexecute\u001b[0;34m(self, query, args)\u001b[0m\n\u001b[1;32m    168\u001b[0m         \u001b[0mquery\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmogrify\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 170\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_query\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    171\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_executed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mquery\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    172\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/cursors.py\u001b[0m in \u001b[0;36m_query\u001b[0;34m(self, q)\u001b[0m\n\u001b[1;32m    326\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_last_executed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m         \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    329\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_do_get_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrowcount\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(self, sql, unbuffered)\u001b[0m\n\u001b[1;32m    515\u001b[0m                 \u001b[0msql\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msql\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'surrogateescape'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    516\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_execute_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCOMMAND\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCOM_QUERY\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msql\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 517\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_affected_rows\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_query_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0munbuffered\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0munbuffered\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    518\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_affected_rows\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    519\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36m_read_query_result\u001b[0;34m(self, unbuffered)\u001b[0m\n\u001b[1;32m    730\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    731\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mMySQLResult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 732\u001b[0;31m             \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    733\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_result\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    734\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mserver_status\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1073\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1074\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1075\u001b[0;31m             \u001b[0mfirst_packet\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_packet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1076\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1077\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mfirst_packet\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_ok_packet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/connections.py\u001b[0m in \u001b[0;36m_read_packet\u001b[0;34m(self, packet_type)\u001b[0m\n\u001b[1;32m    682\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    683\u001b[0m         \u001b[0mpacket\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpacket_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuff\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 684\u001b[0;31m         \u001b[0mpacket\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    685\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mpacket\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/protocol.py\u001b[0m in \u001b[0;36mcheck_error\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    218\u001b[0m             \u001b[0merrno\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_uint16\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    219\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mDEBUG\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"errno =\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrno\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 220\u001b[0;31m             \u001b[0merr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_mysql_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    221\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    222\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdump\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Anaconda/lib/python3.7/site-packages/pymysql/err.py\u001b[0m in \u001b[0;36mraise_mysql_exception\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m    107\u001b[0m         \u001b[0merrval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'utf-8'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'replace'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    108\u001b[0m     \u001b[0merrorclass\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0merror_map\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mInternalError\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 109\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0merrorclass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merrno\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrval\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mIntegrityError\u001b[0m: (1062, \"Duplicate entry 'AT1G01040' for key 'PRIMARY'\")"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv\n",
    "\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)\n",
    "try:    \n",
    "    with connection.cursor() as cursor:\n",
    "        with open (\"LocusGene.tsv\", newline= \"\\n\") as file:\n",
    "            locusgenefile = csv.reader(file, delimiter='\\t', quotechar='\"')\n",
    "            headers = next(locusgenefile) \n",
    "            for row in locusgenefile:\n",
    "                #print(row)\n",
    "                sql = \"INSERT INTO locusgenetab(locus, gene, proteinlengh) VALUES (%s,%s,%s)\" \n",
    "                cursor.execute(sql, tuple(row)) \n",
    "                results = cursor.fetchall()\n",
    "                print(results)       \n",
    "                print(\"\")\n",
    "        \n",
    "finally:\n",
    "    print(\"done!\")\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>locus</th>\n",
       "        <th>gene</th>\n",
       "        <th>proteinlengh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM locusgenetab;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "#secondexam = open('secondexam.txt', 'x') #create the file\n",
    "secondexam = open('secondexam.txt', 'w') #allows to write in the file (not to read)\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)  \n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasmtab, locusgenetab \n",
    "        WHERE germplasmtab.locus = locusgenetab.locus\"\"\" #takes the information of the two tables as long as the column \"locus\" in both are identicals\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        secondexam.write(''.join(str(results)) + '\\n') #write in the \"secondexam\"file the results, by creating an empty string and adding the results\n",
    "        print(\"\")\n",
    "        \n",
    "finally:\n",
    "    print(\"done!\")\n",
    "connection.close() \n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"[{'locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locusgenetab.locus': 'AT1G01040', 'gene': 'DCL1', 'proteinlengh': 332}, {'locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT1G01060', 'gene': 'LHY', 'proteinlengh': 290}, {'locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locusgenetab.locus': 'AT1G01140', 'gene': 'CIPK9', 'proteinlengh': 223}, {'locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locusgenetab.locus': 'AT1G01220', 'gene': 'FKGP', 'proteinlengh': 190}, {'locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locusgenetab.locus': 'AT2G03720', 'gene': 'MRH6', 'proteinlengh': 189}, {'locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locusgenetab.locus': 'AT2G03800', 'gene': 'GEK1', 'proteinlengh': 196}, {'locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locusgenetab.locus': 'AT2G04240', 'gene': 'XERICO', 'proteinlengh': 256}, {'locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locusgenetab.locus': 'AT2G05210', 'gene': 'POT1A', 'proteinlengh': 221}, {'locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locusgenetab.locus': 'AT3G02130', 'gene': 'RPK2', 'proteinlengh': 284}, {'locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locusgenetab.locus': 'AT3G02140', 'gene': 'TMAC2', 'proteinlengh': 300}, {'locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locusgenetab.locus': 'AT3G02230', 'gene': 'RGP1', 'proteinlengh': 301}, {'locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locusgenetab.locus': 'AT3G02260', 'gene': 'BIG', 'proteinlengh': 279}, {'locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locusgenetab.locus': 'AT3G02310', 'gene': 'SEP2', 'proteinlengh': 175}, {'locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locusgenetab.locus': 'AT3G02680', 'gene': 'NBS1', 'proteinlengh': 190}, {'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locusgenetab.locus': 'AT3G02870', 'gene': 'VTC4', 'proteinlengh': 311}, {'locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locusgenetab.locus': 'AT3G03260', 'gene': 'HDG8', 'proteinlengh': 194}, {'locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locusgenetab.locus': 'AT4G14790', 'gene': 'SUV3', 'proteinlengh': 312}, {'locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locusgenetab.locus': 'AT4G15210', 'gene': 'BAM5', 'proteinlengh': 313}, {'locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locusgenetab.locus': 'AT4G15560', 'gene': 'DXS', 'proteinlengh': 219}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}, {'locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locusgenetab.locus': 'AT4G15802', 'gene': 'HSBP', 'proteinlengh': 254}, {'locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locusgenetab.locus': 'AT4G15880', 'gene': 'ESD4', 'proteinlengh': 265}, {'locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locusgenetab.locus': 'AT4G16420', 'gene': 'ADA2B', 'proteinlengh': 279}, {'locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locusgenetab.locus': 'AT4G16480', 'gene': 'INT4', 'proteinlengh': 284}, {'locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locusgenetab.locus': 'AT5G10480', 'gene': 'PAS2', 'proteinlengh': 301}, {'locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locusgenetab.locus': 'AT5G10510', 'gene': 'AIL6', 'proteinlengh': 310}, {'locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locusgenetab.locus': 'AT5G11110', 'gene': 'SPS2', 'proteinlengh': 232}, {'locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT5G11260', 'gene': 'HY5', 'proteinlengh': 221}, {'locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locusgenetab.locus': 'AT5G11510', 'gene': 'MYB3R-4', 'proteinlengh': 336}, {'locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locusgenetab.locus': 'AT5G12200', 'gene': 'PYD2', 'proteinlengh': 310}, {'locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locusgenetab.locus': 'AT5G13290', 'gene': 'CRN', 'proteinlengh': 189}]\\n\"]\n"
     ]
    }
   ],
   "source": [
    "secondexam = open('secondexam.txt', 'r')\n",
    "print(secondexam.readlines()) #to check "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "#Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "import pymysql.cursors\n",
    "secondexam = open('secondexam.txt', 'a')\n",
    "\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)  # note that it is possible to delay putting changes into the database!\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * FROM germplasmtab, locusgenetab \n",
    "        WHERE (locusgenetab.gene = 'SKOR' OR locusgenetab.gene= 'MAA3') AND germplasmtab.locus = locusgenetab.locus\"\"\" #the parentheses are to give priority to the \"or\" because otherwise the \"and\" has priority\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall() \n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "    \n",
    "finally:\n",
    "    print(\"done!\")\n",
    "connection.close() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"[{'locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locusgenetab.locus': 'AT1G01040', 'gene': 'DCL1', 'proteinlengh': 332}, {'locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT1G01060', 'gene': 'LHY', 'proteinlengh': 290}, {'locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locusgenetab.locus': 'AT1G01140', 'gene': 'CIPK9', 'proteinlengh': 223}, {'locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locusgenetab.locus': 'AT1G01220', 'gene': 'FKGP', 'proteinlengh': 190}, {'locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locusgenetab.locus': 'AT2G03720', 'gene': 'MRH6', 'proteinlengh': 189}, {'locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locusgenetab.locus': 'AT2G03800', 'gene': 'GEK1', 'proteinlengh': 196}, {'locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locusgenetab.locus': 'AT2G04240', 'gene': 'XERICO', 'proteinlengh': 256}, {'locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locusgenetab.locus': 'AT2G05210', 'gene': 'POT1A', 'proteinlengh': 221}, {'locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locusgenetab.locus': 'AT3G02130', 'gene': 'RPK2', 'proteinlengh': 284}, {'locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locusgenetab.locus': 'AT3G02140', 'gene': 'TMAC2', 'proteinlengh': 300}, {'locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locusgenetab.locus': 'AT3G02230', 'gene': 'RGP1', 'proteinlengh': 301}, {'locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locusgenetab.locus': 'AT3G02260', 'gene': 'BIG', 'proteinlengh': 279}, {'locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locusgenetab.locus': 'AT3G02310', 'gene': 'SEP2', 'proteinlengh': 175}, {'locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locusgenetab.locus': 'AT3G02680', 'gene': 'NBS1', 'proteinlengh': 190}, {'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locusgenetab.locus': 'AT3G02870', 'gene': 'VTC4', 'proteinlengh': 311}, {'locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locusgenetab.locus': 'AT3G03260', 'gene': 'HDG8', 'proteinlengh': 194}, {'locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locusgenetab.locus': 'AT4G14790', 'gene': 'SUV3', 'proteinlengh': 312}, {'locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locusgenetab.locus': 'AT4G15210', 'gene': 'BAM5', 'proteinlengh': 313}, {'locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locusgenetab.locus': 'AT4G15560', 'gene': 'DXS', 'proteinlengh': 219}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}, {'locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locusgenetab.locus': 'AT4G15802', 'gene': 'HSBP', 'proteinlengh': 254}, {'locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locusgenetab.locus': 'AT4G15880', 'gene': 'ESD4', 'proteinlengh': 265}, {'locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locusgenetab.locus': 'AT4G16420', 'gene': 'ADA2B', 'proteinlengh': 279}, {'locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locusgenetab.locus': 'AT4G16480', 'gene': 'INT4', 'proteinlengh': 284}, {'locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locusgenetab.locus': 'AT5G10480', 'gene': 'PAS2', 'proteinlengh': 301}, {'locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locusgenetab.locus': 'AT5G10510', 'gene': 'AIL6', 'proteinlengh': 310}, {'locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locusgenetab.locus': 'AT5G11110', 'gene': 'SPS2', 'proteinlengh': 232}, {'locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT5G11260', 'gene': 'HY5', 'proteinlengh': 221}, {'locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locusgenetab.locus': 'AT5G11510', 'gene': 'MYB3R-4', 'proteinlengh': 336}, {'locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locusgenetab.locus': 'AT5G12200', 'gene': 'PYD2', 'proteinlengh': 310}, {'locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locusgenetab.locus': 'AT5G13290', 'gene': 'CRN', 'proteinlengh': 189}]\\n\", \"[{'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}]\\n\", \"[{'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}]\\n\"]\n"
     ]
    }
   ],
   "source": [
    "secondexam = open('secondexam.txt', 'r')\n",
    "print(secondexam.readlines())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "#%sql SELECT locus, COUNT(locus) FROM locusgenetab GROUP BY locus HAVING COUNT(locus)>1\n",
    "import re\n",
    "import pymysql.cursors\n",
    "\n",
    "secondexam = open('secondexam.txt', 'a') \n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4', \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)  \n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "       # with open(\"secondexam.txt\", 'w'):\n",
    "        sql = \"SELECT COUNT(locus) FROM germplasmtab WHERE locus REGEXP 'AT1G'\" #COUNT(column) to count how many in that column follows the description, WHERE locus REGEXP 'AT1G' is the description\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\") \n",
    "        sql = \"SELECT COUNT(locus) FROM germplasmtab WHERE locus REGEXP 'AT2G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "        sql = \"SELECT COUNT(locus) FROM germplasmtab WHERE locus REGEXP 'AT3G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "        sql = \"SELECT COUNT(locus) FROM germplasmtab WHERE locus REGEXP 'AT4G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "        sql = \"SELECT COUNT(locus) FROM germplasmtab WHERE locus REGEXP 'AT5G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "    \n",
    "finally:\n",
    "    print(\"done!\")\n",
    "    \n",
    "    \n",
    "connection.close() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"[{'locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locusgenetab.locus': 'AT1G01040', 'gene': 'DCL1', 'proteinlengh': 332}, {'locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT1G01060', 'gene': 'LHY', 'proteinlengh': 290}, {'locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locusgenetab.locus': 'AT1G01140', 'gene': 'CIPK9', 'proteinlengh': 223}, {'locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locusgenetab.locus': 'AT1G01220', 'gene': 'FKGP', 'proteinlengh': 190}, {'locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locusgenetab.locus': 'AT2G03720', 'gene': 'MRH6', 'proteinlengh': 189}, {'locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locusgenetab.locus': 'AT2G03800', 'gene': 'GEK1', 'proteinlengh': 196}, {'locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locusgenetab.locus': 'AT2G04240', 'gene': 'XERICO', 'proteinlengh': 256}, {'locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locusgenetab.locus': 'AT2G05210', 'gene': 'POT1A', 'proteinlengh': 221}, {'locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locusgenetab.locus': 'AT3G02130', 'gene': 'RPK2', 'proteinlengh': 284}, {'locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locusgenetab.locus': 'AT3G02140', 'gene': 'TMAC2', 'proteinlengh': 300}, {'locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locusgenetab.locus': 'AT3G02230', 'gene': 'RGP1', 'proteinlengh': 301}, {'locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locusgenetab.locus': 'AT3G02260', 'gene': 'BIG', 'proteinlengh': 279}, {'locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locusgenetab.locus': 'AT3G02310', 'gene': 'SEP2', 'proteinlengh': 175}, {'locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locusgenetab.locus': 'AT3G02680', 'gene': 'NBS1', 'proteinlengh': 190}, {'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locusgenetab.locus': 'AT3G02870', 'gene': 'VTC4', 'proteinlengh': 311}, {'locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locusgenetab.locus': 'AT3G03260', 'gene': 'HDG8', 'proteinlengh': 194}, {'locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locusgenetab.locus': 'AT4G14790', 'gene': 'SUV3', 'proteinlengh': 312}, {'locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locusgenetab.locus': 'AT4G15210', 'gene': 'BAM5', 'proteinlengh': 313}, {'locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locusgenetab.locus': 'AT4G15560', 'gene': 'DXS', 'proteinlengh': 219}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}, {'locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locusgenetab.locus': 'AT4G15802', 'gene': 'HSBP', 'proteinlengh': 254}, {'locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locusgenetab.locus': 'AT4G15880', 'gene': 'ESD4', 'proteinlengh': 265}, {'locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locusgenetab.locus': 'AT4G16420', 'gene': 'ADA2B', 'proteinlengh': 279}, {'locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locusgenetab.locus': 'AT4G16480', 'gene': 'INT4', 'proteinlengh': 284}, {'locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locusgenetab.locus': 'AT5G10480', 'gene': 'PAS2', 'proteinlengh': 301}, {'locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locusgenetab.locus': 'AT5G10510', 'gene': 'AIL6', 'proteinlengh': 310}, {'locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locusgenetab.locus': 'AT5G11110', 'gene': 'SPS2', 'proteinlengh': 232}, {'locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT5G11260', 'gene': 'HY5', 'proteinlengh': 221}, {'locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locusgenetab.locus': 'AT5G11510', 'gene': 'MYB3R-4', 'proteinlengh': 336}, {'locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locusgenetab.locus': 'AT5G12200', 'gene': 'PYD2', 'proteinlengh': 310}, {'locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locusgenetab.locus': 'AT5G13290', 'gene': 'CRN', 'proteinlengh': 189}]\\n\", \"[{'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}]\\n\", \"[{'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}]\\n\", \"[{'COUNT(locus)': 4}]\\n\", \"[{'COUNT(locus)': 4}]\\n\", \"[{'COUNT(locus)': 9}]\\n\", \"[{'COUNT(locus)': 8}]\\n\", \"[{'COUNT(locus)': 7}]\\n\"]\n"
     ]
    }
   ],
   "source": [
    "secondexam = open('secondexam.txt', 'r')\n",
    "print(secondexam.readlines())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "secondexam = open('secondexam.txt', 'a')\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='genetics',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "connection.autocommit(True)  \n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT AVG(proteinlengh) FROM locusgenetab WHERE locus REGEXP 'AT1G'\" #the same but to get the average\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\") \n",
    "        sql = \"SELECT AVG(proteinlengh) FROM locusgenetab WHERE locus REGEXP 'AT2G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "        sql = \"SELECT AVG(proteinlengh) FROM locusgenetab WHERE locus REGEXP 'AT3G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "        sql = \"SELECT AVG(proteinlengh) FROM locusgenetab WHERE locus REGEXP 'AT4G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\") \n",
    "        sql = \"SELECT AVG(proteinlengh) FROM locusgenetab WHERE locus REGEXP 'AT5G'\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall()\n",
    "        for t in results:\n",
    "            secondexam.write(''.join(str(results)) + '\\n') \n",
    "        print(\"\")\n",
    "    \n",
    "finally:\n",
    "    print(\"done!\")\n",
    "    \n",
    "    \n",
    "connection.close() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"[{'locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locusgenetab.locus': 'AT1G01040', 'gene': 'DCL1', 'proteinlengh': 332}, {'locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT1G01060', 'gene': 'LHY', 'proteinlengh': 290}, {'locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locusgenetab.locus': 'AT1G01140', 'gene': 'CIPK9', 'proteinlengh': 223}, {'locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locusgenetab.locus': 'AT1G01220', 'gene': 'FKGP', 'proteinlengh': 190}, {'locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locusgenetab.locus': 'AT2G03720', 'gene': 'MRH6', 'proteinlengh': 189}, {'locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locusgenetab.locus': 'AT2G03800', 'gene': 'GEK1', 'proteinlengh': 196}, {'locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locusgenetab.locus': 'AT2G04240', 'gene': 'XERICO', 'proteinlengh': 256}, {'locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locusgenetab.locus': 'AT2G05210', 'gene': 'POT1A', 'proteinlengh': 221}, {'locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locusgenetab.locus': 'AT3G02130', 'gene': 'RPK2', 'proteinlengh': 284}, {'locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locusgenetab.locus': 'AT3G02140', 'gene': 'TMAC2', 'proteinlengh': 300}, {'locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locusgenetab.locus': 'AT3G02230', 'gene': 'RGP1', 'proteinlengh': 301}, {'locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locusgenetab.locus': 'AT3G02260', 'gene': 'BIG', 'proteinlengh': 279}, {'locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locusgenetab.locus': 'AT3G02310', 'gene': 'SEP2', 'proteinlengh': 175}, {'locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locusgenetab.locus': 'AT3G02680', 'gene': 'NBS1', 'proteinlengh': 190}, {'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locusgenetab.locus': 'AT3G02870', 'gene': 'VTC4', 'proteinlengh': 311}, {'locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locusgenetab.locus': 'AT3G03260', 'gene': 'HDG8', 'proteinlengh': 194}, {'locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locusgenetab.locus': 'AT4G14790', 'gene': 'SUV3', 'proteinlengh': 312}, {'locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locusgenetab.locus': 'AT4G15210', 'gene': 'BAM5', 'proteinlengh': 313}, {'locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locusgenetab.locus': 'AT4G15560', 'gene': 'DXS', 'proteinlengh': 219}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}, {'locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locusgenetab.locus': 'AT4G15802', 'gene': 'HSBP', 'proteinlengh': 254}, {'locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locusgenetab.locus': 'AT4G15880', 'gene': 'ESD4', 'proteinlengh': 265}, {'locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locusgenetab.locus': 'AT4G16420', 'gene': 'ADA2B', 'proteinlengh': 279}, {'locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locusgenetab.locus': 'AT4G16480', 'gene': 'INT4', 'proteinlengh': 284}, {'locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locusgenetab.locus': 'AT5G10480', 'gene': 'PAS2', 'proteinlengh': 301}, {'locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locusgenetab.locus': 'AT5G10510', 'gene': 'AIL6', 'proteinlengh': 310}, {'locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locusgenetab.locus': 'AT5G11110', 'gene': 'SPS2', 'proteinlengh': 232}, {'locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locusgenetab.locus': 'AT5G11260', 'gene': 'HY5', 'proteinlengh': 221}, {'locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locusgenetab.locus': 'AT5G11510', 'gene': 'MYB3R-4', 'proteinlengh': 336}, {'locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locusgenetab.locus': 'AT5G12200', 'gene': 'PYD2', 'proteinlengh': 310}, {'locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locusgenetab.locus': 'AT5G13290', 'gene': 'CRN', 'proteinlengh': 189}]\\n\", \"[{'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}]\\n\", \"[{'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgenetab.locus': 'AT3G02850', 'gene': 'SKOR', 'proteinlengh': 234}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgenetab.locus': 'AT4G15570', 'gene': 'MAA3', 'proteinlengh': 294}]\\n\", \"[{'COUNT(locus)': 4}]\\n\", \"[{'COUNT(locus)': 4}]\\n\", \"[{'COUNT(locus)': 9}]\\n\", \"[{'COUNT(locus)': 8}]\\n\", \"[{'COUNT(locus)': 7}]\\n\", \"[{'AVG(proteinlengh)': Decimal('258.7500')}]\\n\", \"[{'AVG(proteinlengh)': Decimal('215.5000')}]\\n\", \"[{'AVG(proteinlengh)': Decimal('252.0000')}]\\n\", \"[{'AVG(proteinlengh)': Decimal('277.5000')}]\\n\", \"[{'AVG(proteinlengh)': Decimal('271.2857')}]\\n\"]\n"
     ]
    }
   ],
   "source": [
    "secondexam = open('secondexam.txt', 'r')\n",
    "print(secondexam.readlines())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
